File: 10_Binary_search_2D_array/Binary_search.java
class Binary_search {
    
    public boolean searchMatrix(int[][] matrix, int target) {
        int N = matrix.length;      
        int M = matrix[0].length;   
        int row_index, col_index;
        int start = 0, end = N * M - 1, mid;

        
        while (start <= end) {
            mid = start + (end - start) / 2;
            row_index = mid / M;       // Calculate the row index
            col_index = mid % M;       // Calculate the column index

            if (matrix[row_index][col_index] == target) {
                return true;
            }
            else if (matrix[row_index][col_index] < target) {
                start = mid + 1;
            }
            else {
                end = mid - 1;
            }
        }

        return false;
    }

    
    public static void main(String[] args) {
        Binary_search solution = new Binary_search();

        // Test matrix
        int[][] matrix = {
            {1, 3, 5, 7},
            {10, 11, 16, 20},
            {23, 30, 34, 60}
        };

        // Test case 1: Searching for target 3 (should return true)
        int target1 = 3;
        System.out.println("Search for " + target1 + ": " + solution.searchMatrix(matrix, target1));

        // Test case 2: Searching for target 13 (should return false)
        int target2 = 13;
        System.out.println("Search for " + target2 + ": " + solution.searchMatrix(matrix, target2));

        // Test case 3: Searching for target 23 (should return true)
        int target3 = 23;
        System.out.println("Search for " + target3 + ": " + solution.searchMatrix(matrix, target3));
    }
}

File: 10_Binary_search_2D_array/Binary_search.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int N = matrix.size(), M = matrix[0].size();
        int row_index, col_index, start = 0, end = N*M-1, mid;

        while(start <= end){
            mid = start+(end-start)/2;
            row_index = mid/M;
            col_index = mid%M;

            if(matrix[row_index][col_index] == target) return true;
            else if (matrix[row_index][col_index] < target) start = mid+1;
            else end = mid-1;
        }

        return false;
    }
};

int main(){
    Solution solution;

    vector<vector<int>> matrix = {
        {1, 4, 7, 11},
        {12, 15, 18, 19},
        {20, 22, 25, 27},
        {30, 32, 35, 38}
    };

    int target = 11;

    if (solution.searchMatrix(matrix, target)){
        cout << "true" << endl;
    } else{
        cout << "false" << endl;
    }

    return 0;
}